invokeai/backend/model_management2/model_config.py

```diff
@@ -0,0 +1,276 @@
+  print(config.name)
+
+Validation errors will raise an InvalidModelConfigException error.
+
+"""
+import pydantic
+from enum import Enum
+from pydantic import BaseModel, Field, Extra
+from typing import Optional, Literal, List, Union, Type
+from pydantic.error_wrappers import ValidationError
+from omegaconf.listconfig import ListConfig  # to support the yaml backend
+
+
+class InvalidModelConfigException(Exception):
+    """Exception raised when the config parser doesn't recognize the passed
+    combination of model type and format."""
+
+    pass
+
+
+class BaseModelType(str, Enum):
+    """Base model type."""
+
+    StableDiffusion1 = "sd-1"
+    StableDiffusion2 = "sd-2"
+    StableDiffusionXL = "sdxl"
+    StableDiffusionXLRefiner = "sdxl-refiner"
+    # Kandinsky2_1 = "kandinsky-2.1"
+
+
+class ModelType(str, Enum):
+    """Model type."""
+
+    ONNX = "onnx"
+    Main = "main"
+    Vae = "vae"
+    Lora = "lora"
+    ControlNet = "controlnet"  # used by model_probe
+    TextualInversion = "embedding"
+
+
+class SubModelType(str, Enum):
+    """Submodel type."""
+
+    UNet = "unet"
+    TextEncoder = "text_encoder"
+    TextEncoder2 = "text_encoder_2"
+    Tokenizer = "tokenizer"
+    Tokenizer2 = "tokenizer_2"
+    Vae = "vae"
+    VaeDecoder = "vae_decoder"
+    VaeEncoder = "vae_encoder"
+    Scheduler = "scheduler"
+    SafetyChecker = "safety_checker"
+
+
+class ModelVariantType(str, Enum):
+    """Variant type."""
+
+    Normal = "normal"
+    Inpaint = "inpaint"
+    Depth = "depth"
+
+
+class ModelFormat(str, Enum):
+    """Storage format of model."""
+
+    Diffusers = "diffusers"
+    Checkpoint = "checkpoint"
+    Lycoris = "lycoris"
+    Onnx = "onnx"
+    Olive = "olive"
+    EmbeddingFile = "embedding_file"
+    EmbeddingFolder = "embedding_folder"
+
+
+class SchedulerPredictionType(str, Enum):
+    """Scheduler prediction type."""
+
+    Epsilon = "epsilon"
+    VPrediction = "v_prediction"
+    Sample = "sample"
+
+
+class ModelConfigBase(BaseModel):
+    """Base class for model configuration information."""
+
+    path: str
+    name: str
+    base_model: BaseModelType
+    model_type: ModelType
+    model_format: ModelFormat
+    id: Optional[str] = Field(None)  # this may get added by the store
+    description: Optional[str] = Field(None)
+    author: Optional[str] = Field(description="Model author")
+    thumbnail_url: Optional[str] = Field(description="URL of thumbnail image")
+    license_url: Optional[str] = Field(description="URL of license")
+    source_url: Optional[str] = Field(description="Model download source")
+    tags: Optional[List[str]] = Field(description="Descriptive tags")  # Set would be better, but not JSON serializable
+
+    class Config:
+        """Pydantic configuration hint."""
+
+        use_enum_values = True
+        extra = Extra.forbid
+        validate_assignment = True
+
+    @pydantic.validator("tags", pre=True)
+    @classmethod
+    def _fix_tags(cls, v):
```

Comment:
Commenting on my own code here - I don't like the fact that encapsulation is broken here in order to accommodate a quirk of the YAML storage backend.